src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -351,22 +351,20 @@ EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index,
 }
 
 bool
-EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, size_t count, char*& characters)
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, size_t count, dynamic_local_string<SENSIBLE_PATH_MAX>& characters)
 {
 	if (!zip_ensure_valid_params (buf, buf_len, index, count)) {
 		return false;
 	}
 
-	size_t alloc_size = ADD_WITH_OVERFLOW_CHECK (size_t, count, 1);
-	characters = new char[alloc_size];
-	memcpy (characters, buf + index, count);
-	characters [count] = '\0';
+	memcpy (characters.get (), buf + index, count);
```

Comment:
This feels unsafe without some form of check that `count` < `SENSIBLE_PATH_MAX`.  Quick codepath check, and I see that in `zip_read_entry_info()`, `zip_read_field()` is called with `count` coming from `file_name_length`, a `uint16_t`.  It is thus *plausible*, without further checking, that `count` could be `MAX_SHORT`, or 64k, which can be longer than `SENSIBLE_PATH_MAX`.

I think instead there should be a `dynamic_local_string::assign(char* source, size_t offset, size_t count)` method, which can appropriately reallocate the underlying storage if necessary.

Reply:
I see that there's already a `dynamic_local_string::append()` member function.  Why not use that instead?

```cpp
characters.clear ();
characters.assign (buf + index, count);
```